File: src/statement/block.ts
import { Pass } from "../pass/pass";
import { SymbolTable } from "../symbols/symbolTable";
import { Node } from "./node";
import { Statement } from "./statement";

export class Block implements Node {
    public symbolTable = new SymbolTable();

    constructor(public statements: Statement[], public parent: Block | null) {}

    public accept(pass: Pass): void {
        pass.visitBlock(this);
    }

    public getSymbolByName(name: string): any {
        if (this.symbolTable.has(name))
            return this.symbolTable.get(name);

        if (this.parent)
            return this.parent.getSymbolByName(name);

        return null;
    }
}

File: common/src/main/kotlin/me/honkling/skriptkt/common/node/Block.kt
package me.honkling.skriptkt.common.node

import me.honkling.skriptkt.common.statement.Statement

class Block(parent: Block?, val statements: MutableList<Statement>) : Node(parent) {
    override fun accept() {
        for (statement in statements)
            statement.accept()
    }

    override fun toString(): String {
        return "Block(statements=$statements)"
    }
}